File: lc_cpp/lcCpp/BitCalculation/singleNumber.c
#include <stdio.h>
#include <assert.h>

int singleNumber(int* nums, int numsSize) {
    int ans = 0;
    for (int i = 0; i < 32; i++) {
        int cnt1 = 0;
        for (int j = 0; j < numsSize; j++) {
            cnt1 += (nums[j] >> i) & 1;
        }
        ans |= (cnt1 % 3) << i;
    }
    return ans;
}

// 状态机解法
int singleNumberStateMachine(int* nums, int numsSize) {
    int ones = 0, twos = 0;
    for (int i = 0; i < numsSize; i++) {
        ones = (ones ^ nums[i]) & ~twos;
        twos = (twos ^ nums[i]) & ~ones;
    }
    return ones;
}

void test() {
    int nums1[] = {2, 2, 3, 2};
    assert(singleNumber(nums1, 4) == 3);
    
    int nums2[] = {0, 1, 0, 1, 0, 1, 99};
    assert(singleNumber(nums2, 7) == 99);
    
    printf("All tests passed!\n");
}

int main() {
    test();
    return 0;
}

File: lc_ts/lcTs/bitCalculation/singleNumber.ts
class Solution {
    // 位运算解法
    singleNumber(nums: number[]): number {
        let ans = 0;
        for (let i = 0; i < 32; i++) {
            let cnt1 = 0;
            for (const x of nums) {
                cnt1 += (x >> i) & 1;
            }
            ans |= (cnt1 % 3) << i;
        }
        return ans;
    }
    
    // 状态机解法
    singleNumberStateMachine(nums: number[]): number {
        let ones = 0, twos = 0;
        for (const num of nums) {
            ones = (ones ^ num) & ~twos;
            twos = (twos ^ num) & ~ones;
        }
        return ones;
    }
    
    // 泛型解法
    singleNumberGeneric<T extends number>(nums: T[]): T {
        let ans = 0 as T;
        for (let i = 0; i < 32; i++) {
            let cnt1 = 0;
            for (const x of nums) {
                cnt1 += (x >> i) & 1;
            }
            ans |= ((cnt1 % 3) << i) as T;
        }
        return ans;
    }
}

// 函数式编程解法
const singleNumberFunctional = (nums: number[]): number => {
    return Array.from({length: 32}, (_, i) => i)
        .reduce((ans, i) => {
            const cnt1 = nums.reduce((count, x) => count + ((x >> i) & 1), 0);
            return ans | ((cnt1 % 3) << i);
        }, 0);
};

// 测试
function test(): void {
    const sol = new Solution();
    console.assert(sol.singleNumber([2, 2, 3, 2]) === 3);
    console.assert(sol.singleNumber([0, 1, 0, 1, 0, 1, 99]) === 99);
    console.log("All tests passed!");
}

test();